source/secure_tunneling_operations.c

```diff
@@ -0,0 +1,670 @@
+    aws_ref_count_init(&message_op->base.ref_count, message_op, s_destroy_operation_message);
+    message_op->base.impl = message_op;
+
+    if (aws_secure_tunnel_message_storage_init(&message_op->options_storage, allocator, message_options, type)) {
+        goto error;
+    }
+
+    message_op->base.message_view = &message_op->options_storage.storage_view;
+
+    return message_op;
+
+error:
+
+    aws_secure_tunnel_operation_release(&message_op->base);
+
+    return NULL;
+}
+
+/*********************************************************************************************************************
+ * Pingreq
+ ********************************************************************************************************************/
+
+static void s_destroy_operation_pingreq(void *object) {
+    if (object == NULL) {
+        return;
+    }
+
+    struct aws_secure_tunnel_operation_pingreq *pingreq_op = object;
+    aws_mem_release(pingreq_op->allocator, pingreq_op);
+}
+
+struct aws_secure_tunnel_operation_pingreq *aws_secure_tunnel_operation_pingreq_new(struct aws_allocator *allocator) {
+    AWS_PRECONDITION(allocator != NULL);
+
+    struct aws_secure_tunnel_operation_pingreq *pingreq_op =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_secure_tunnel_operation_pingreq));
+    if (pingreq_op == NULL) {
+        return NULL;
+    }
+
+    pingreq_op->allocator = allocator;
+    pingreq_op->base.vtable = &s_empty_operation_vtable;
+    pingreq_op->base.operation_type = AWS_STOT_PING;
+    aws_ref_count_init(&pingreq_op->base.ref_count, pingreq_op, s_destroy_operation_pingreq);
+    pingreq_op->base.impl = pingreq_op;
+
+    return pingreq_op;
+}
+
+/*********************************************************************************************************************
+ * Secure Tunnel Storage Options
+ ********************************************************************************************************************/
+
+/*
+ * Validation of options on creation of a new secure tunnel
+ */
+int aws_secure_tunnel_options_validate(const struct aws_secure_tunnel_options *options) {
+    AWS_ASSERT(options && options->allocator);
+    if (options->bootstrap == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "bootstrap cannot be NULL");
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+    if (options->socket_options == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "socket options cannot be NULL");
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+    if (options->access_token.len == 0) {
+        AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "access token is required");
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+    if (options->endpoint_host.len == 0) {
+        AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "endpoint host is required");
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
```

Comment:
Are there any constraints on service ids?  Can 2 be defined but not 1 for example?

Reply:
0-3 service ids can be set and as far as I've seen, they're not numbered ordered in a way that can skip an index. I should probably rework some of the logic checking service ids to take this into account. Though shifting to a hash table would just do that automatically.